### ISSUE TITLE
Feature Request: Add filtering to delete All

### ISSUE THREAD

**Issue by ZMomo** [vuex-orm/vuex-orm#345]
Hey,

I think it would be interesting to add a feature to the deleteAll method to delete filtered data.

ex: User.DeleteAll(1) would delete all data from User with id 1, for now we can either delete data from all Users or delete specific users but not all their associated entities.

It could be nice to implement an instance method $deleteAll()
exemple: 
let user = User.find(1)
user.$deleteAll()       //would delete all associated data from user with id 1

What do you think ? I am trying to implement integrity constraints like cascade deleting and i think this feature would make it easier.

**Comment by kiaking** [comment]
@ZMomo Hi! Thank you for the feedback!

OK, so I think what you want is a way to delete any relationship associated to the entity. Yeah, that might be nice. But I would like to discuss about the API.

Since there is already `deleteAll` method, I think we should have different name, or either accept option. An entity could have bunch of relationship, so I think it's more safe to let user to specify which relationship to delete too. Maybe something like;

```js
// Delete user with id of 1 and any posts associates with the user.
User.delete(1, {
  cascade: ['posts']
})

// Delete user with id of 1 and all relationships associates with the user.
User.delete(1, {
  cascade: ['*']
})
```

What do you think?

**Comment by ZMomo** [comment]
Hi @kiaking ,

I think it's even more convenient this way ! I think it would be awsome ! 

Thanks for fast response.

**Comment by kiaking** [comment]
@ZMomo Great! Thank you for the confirmation. There're other issues I have to tackle so it might take awhile to implement this, but I think it's nice feature to have! Meanwhile maybe someone can tackle this one. 🤝

**Comment by ZMomo** [comment]
Maybe it would be nice to have a bit more complexe querying:

`// Delete user with id of 1 and any posts associates with the user.
User.delete(1, {
  cascade: ['posts']
})

// Delete user with id of 1 and all relationships associates with the user.
User.delete(1, {
  cascade: ['*']
})

// Delete user where id 1 and all relationships associates where entitiy is not post.
User.delete(1, {
  cascade: entity => entity != 'posts'
})`

**Comment by miraclebg** [comment]
I am looking at this issue right now after trying to find out why related objects do not get deleted.

I've searched in the documentation but it is not stated anywhere that this would not happen by default (and my presumption would be that it should happen - alike a relationship database).

So could anyone confirm - do the relationships currently get deleted with the deletion of the parent object or not?

I can help out with this once i have a spare moment if it is not implemented.

Thanks!

**Comment by miraclebg** [comment]
More specifically i am looking for a way to do this:

```
MemberMeasurementPhoto.query()
                    .where("memberId", this.id)
                    .where("isLatestLog", true)
                   .delete();
```

Basically a filtered deletion - but in depth of all related objects.

**Comment by kiaking** [comment]
> So could anyone confirm - do the relationships currently get deleted with the deletion of the parent object or not?

No, the relationship will not be deleted.

> I can help out with this once i have a spare moment if it is not implemented.

That is awesome! I think we can use same syntax for instance method as well.

```js
MemberMeasurementPhoto.query()
  .where("memberId", this.id)
  .where("isLatestLog", true)
  .delete({ cascade: ['*'] });
```

**Comment by zzachattack2** [comment]
Would it make more sense to define cascading behavior on the relationship itself? Seems that is pattern most backend ORMs use.

**Comment by kiaking** [comment]
@zzachattack Yeah that would be an idea too. If we can get a cascading feature implemented, we could surely add global (relationship definition) based cascading as well.

Let's don't forget to add that issue after completing this issue : )

**Comment by JonArnfred** [comment]
I'm looking into vuex-orm exactly because cascading deletes frontend is pretty tedious to do by hand, so would love this feature to be implemented. As most apps probably already have defined cascades in the backend database on the relationship, I agree this is the right place to put it. Maybe we could even have the full capacity for the usual on delete handlers:

```javascript

user: this.belongsTo(User, 'user_id')->onDelete('cascade'); // cascade
user: this.belongsTo(User, 'user_id')->onDelete('set null'); // set null
user: this.belongsTo(User, 'user_id')->onDelete('restrict'); // restrict

```

Regardless, thank you for a great library, looking forward to migrate from my own "database" implementation frontend to this one 👍

**Comment by Tofandel** [comment]
I think it would be quite simple to just add a few method to relationships

Like:

```js
comment: this.hasMany(Comment, 'user_id').cascadeOnDelete()
post: this.hasMany(Post, 'user_id').setNullOnDelete()
Address: this.hasOne(Address, 'user_id').restrictOnDelete()
```

And then use the beforeDelete and afterDelete hooks to trigger the relations to also delete, update or abort the deletion

For relationships with pivot (eg: belongsToMany) those methods would only apply to the pivot relation


Currently this can be implemented in userland with


```js
class User extends Model {
  // ...
  static beforeDelete(model) {
    // Restrict on delete
    if (Address.query().where('user_id', model[this.primaryKey]).count()) {
        throw new Exception('restrictOnDelete')
    }
  }
  
  static afterDelete(model) {
    // Set null on delete
    const post = Post.update({
      where: (p) => p.user_id === model[this.primaryKey],
      data: { user_id: null },
    });
    // Cascade on delete
    Comment.delete((c) => c.user_id === model[this.primaryKey]);
  }
}
```
